lib/seahorn/BvOpsem2WideMemMgr.cc

```diff
@@ -0,0 +1,543 @@
+    Expr toExpr() const { return v(); }
+    explicit operator Expr() const { return toExpr(); }
+  };
+
+  using PtrTy = PtrTyImpl;
+  using PtrSortTy = PtrSortTyImpl;
+  using MemValTy = MemValTyImpl;
+  using MemSortTy = MemSortTyImpl;
+
+  /// \brief A null pointer expression (cache)
+  PtrTy m_nullPtr;
+
+  WideMemManager(Bv2OpSem &sem, Bv2OpSemContext &ctx, unsigned ptrSz,
+                 unsigned wordSz, bool useLambdas = false);
+
+  ~WideMemManager() = default;
+
+  Expr bytesToSlotExpr(unsigned bytes) {
+    return m_ctx.alu().si(8 * bytes, g_slotBitWidth);
+  }
+
+  PtrSortTy ptrSort() const {
+    return PtrSortTy(OpSemMemManager::PtrSortTy(),
+                     m_ctx.alu().intTy(g_slotBitWidth));
+  }
+
+  PtrTy salloc(unsigned int bytes, uint32_t align) {
+    assert(isa<AllocaInst>(m_ctx.getCurrentInst()));
+    align = std::max(align, m_alignment);
+    auto region = m_allocator->salloc(bytes, align);
+    assert(region.first - region.second > 0);
+    return PtrTy(mkStackPtr(region.second).getRaw(),
+                 bytesToSlotExpr(region.first - region.second));
+  }
+
+  PtrTy salloc(Expr elmts, unsigned int typeSz, uint32_t align) {
+    align = std::max(align, m_alignment);
+
+    // -- compute number of bytes needed
+    Expr bytes = elmts;
+    if (typeSz > 1) {
+      // TODO: factor out multiplication and number creation
+      bytes = m_ctx.alu().doMul(bytes, m_ctx.alu().si(typeSz, ptrSzInBits()),
+                                ptrSzInBits());
+    }
+
+    // allocate
+    auto region = m_allocator->salloc(bytes, align);
+
+    // -- if allocation failed, return some pointer
+    if (m_allocator->isBadAddrInterval(region)) {
+      LOG("opsem", WARN << "imprecise handling of dynamically "
+                        << "sized stack allocation of " << *elmts
+                        << " elements of size" << typeSz << " bytes\n";);
+      return freshPtr();
+    }
+
+    // -- have a good region, return pointer to it
+    return PtrTy(
+        mkStackPtr(region.second).getRaw(),
+        m_ctx.alu().si(8 * (region.first - region.second), g_slotBitWidth));
+  }
+
+  PtrTy mkStackPtr(unsigned int offset) {
+    return PtrTy(m_main.mkStackPtr(offset), m_uninit_size);
+  }
+
+  PtrTy brk0Ptr() { return PtrTy(m_main.brk0Ptr(), m_uninit_size); }
+
+  PtrTy halloc(unsigned int _bytes, uint32_t align) {
+    PtrTy res = freshPtr();
+    LOG("opsem", WARN << "halloc() not implemented!\n");
+    return res;
+  }
+
+  PtrTy halloc(Expr bytes, uint32_t align) {
+    PtrTy res = freshPtr();
+    LOG("opsem", WARN << "halloc() not implemented!\n");
+    return res;
+  }
+
+  PtrTy galloc(const GlobalVariable &gv, uint32_t align) {
+    uint64_t gvSz = m_sem.getTD().getTypeAllocSize(gv.getValueType());
+    auto range = m_allocator->galloc(gv, gvSz, std::max(align, m_alignment));
+    return PtrTy(m_ctx.alu().si(range.first, ptrSzInBits()),
+                 bytesToSlotExpr(range.second - range.first));
```

Comment:
maybe `range` should have `size()` method so that we don't have to do this computation each time

Reply:
AddrInterval is a std::pair... i can
1. add a size() fn in bvopsem2context OR
2. change AddrInterval to be a class with a size() operator OR
3. probably add a template method -- will have to figure out how...

Reply:
make it into a proper class. It is a pair because it did not need additional methods before